File: repos/openrappter/python/openrappter/channels/registry.py
from openrappter.channels.base import OutgoingMessage


class ChannelRegistry:
    def __init__(self):
        self._channels = {}  # name -> BaseChannel

    def register(self, channel):
        self._channels[channel.name] = channel

    def unregister(self, name):
        self._channels.pop(name, None)

    def has(self, name):
        return name in self._channels

    def get(self, name):
        return self._channels.get(name)

    def names(self):
        return list(self._channels.keys())

    def list(self):
        return list(self._channels.values())

    @property
    def size(self):
        return len(self._channels)

    def connect_all(self):
        for ch in self._channels.values():
            ch.connect()

    def disconnect_all(self):
        for ch in self._channels.values():
            ch.disconnect()

    def send_message(self, channel_id, conversation_id, content):
        ch = self._channels.get(channel_id)
        if not ch:
            raise ValueError(f'Channel not found: {channel_id}')
        msg = OutgoingMessage(
            channel_id=channel_id,
            conversation_id=conversation_id,
            content=content,
        )
        ch.send(conversation_id, msg)
        return msg

    def get_status_list(self):
        return [ch.get_status() for ch in self._channels.values()]

File: repos/rappterbook-agent/typescript/src/channels/registry.ts
/**
 * Channel registry - manages channel instances
 */

import type { BaseChannel } from './base.js';

export interface ChannelStatusInfo {
  id: string;
  type: string;
  connected: boolean;
  configured: boolean;
  running: boolean;
  lastActivity?: string;
  lastConnectedAt?: string;
  lastError?: string;
  messageCount: number;
}

export class ChannelRegistry {
  private channels: Map<string, BaseChannel> = new Map();

  register(channel: BaseChannel): void {
    this.channels.set(channel.name, channel);
  }

  unregister(name: string): boolean {
    return this.channels.delete(name);
  }

  get(name: string): BaseChannel | undefined {
    return this.channels.get(name);
  }

  has(name: string): boolean {
    return this.channels.has(name);
  }

  list(): BaseChannel[] {
    return Array.from(this.channels.values());
  }

  names(): string[] {
    return Array.from(this.channels.keys());
  }

  async connectAll(): Promise<void> {
    const promises = this.list().map(ch => ch.connect());
    await Promise.all(promises);
  }

  async disconnectAll(): Promise<void> {
    const promises = this.list().map(ch => ch.disconnect());
    await Promise.all(promises);
  }

  async connectChannel(type: string): Promise<void> {
    const ch = this.findByType(type);
    if (!ch) throw new Error(`Channel ${type} not registered`);
    await ch.connect();
  }

  async disconnectChannel(type: string): Promise<void> {
    const ch = this.findByType(type);
    if (!ch) throw new Error(`Channel ${type} not registered`);
    await ch.disconnect();
  }

  async probeChannel(type: string): Promise<{ ok: boolean; error?: string }> {
    const ch = this.findByType(type);
    if (!ch) return { ok: false, error: `Channel ${type} not registered` };
    return { ok: ch.connected, error: ch.connected ? undefined : 'Not connected' };
  }

  configureChannel(type: string, config: Record<string, unknown>): void {
    const ch = this.findByType(type);
    if (!ch) throw new Error(`Channel ${type} not registered`);
    ch.setConfig(config);
  }

  getChannelConfig(type: string): { config: Record<string, unknown>; fields: { key: string; label: string; type: string; required: boolean }[] } {
    const ch = this.findByType(type);
    if (!ch) throw new Error(`Channel ${type} not registered`);
    return { config: ch.getConfig(), fields: ch.getConfigFields() };
  }

  async sendMessage(request: { channelId: string; conversationId: string; content: string }): Promise<void> {
    const ch = this.findByType(request.channelId) ?? this.get(request.channelId);
    if (!ch) throw new Error(`Channel ${request.channelId} not found`);
    await ch.send(request.conversationId, {
      channel: request.channelId,
      content: request.content,
    });
  }

  getStatusList(): ChannelStatusInfo[] {
    return this.list().map(ch => {
      const info = ch.getInfo();
      return {
        id: info.name,
        type: info.type,
        connected: info.status === 'connected',
        configured: true,
        running: info.status === 'connected' || info.status === 'connecting',
        lastActivity: info.connectedAt,
        lastConnectedAt: info.connectedAt,
        messageCount: info.messageCount,
      };
    });
  }

  private findByType(type: string): BaseChannel | undefined {
    for (const ch of this.channels.values()) {
      if (ch.type === type || ch.name === type) return ch;
    }
    return undefined;
  }

  clear(): void {
    this.channels.clear();
  }

  get size(): number {
    return this.channels.size;
  }
}
